bitcoin/src/taproot.rs

```diff
@@ -442,6 +464,37 @@ impl TaprootBuilder {
     /// Checks if the builder has finalized building a tree.
     pub fn is_finalizable(&self) -> bool { self.branch.len() == 1 && self.branch[0].is_some() }
 
+    /// Converts the builder into a [`NodeInfo`] if the builder is a full tree with possibly
+    /// hidden nodes
+    ///
+    /// # Errors:
+    ///
+    /// IncompleteBuilder::NotFinalized if the builder is not finalized. The builder
+    /// can be restored by calling [`IncompleteBuilder::into_builder`]
+    pub fn try_into_node_info(mut self) -> Result<NodeInfo, IncompleteBuilder> {
+        if self.branch().len() != 1 {
+            return Err(IncompleteBuilder::NotFinalized(self));
+        }
+        if let Some(Some(node)) = self.branch.pop() {
+            Ok(node)
+        } else {
+            // It is annoying to get rid of this unreachable while maintaining the gurantee that
+            // builder should not change if there is an error
+            unreachable!("Size checked above. Builder guarantees the last element is Some")
```

Comment:
For the future: could we factor out the last element (store it separately), so that the guarantee is represented in type?

Reply:
Great idea.

Reply:
Indeed. I think this should be possible.